Meadowlark-js/packages/meadowlark-core/src/handler/NaturalKeyExtractor.ts

```diff
@@ -40,109 +41,114 @@ export type NaturalKeyWithSecurity = {
+ * Substitute an assignableIdentity from an already constructed DocumentIdentity, if the entity should have one.
+ * If the entity is a subclass with an identity rename, replace the renamed identity property with the original
+ * superclass identity property name, thereby putting it in superclass form.
  *
- * Simple substitution can break down if there are ordering differences in natural key components,
- * however the DS has never had subclasses with multiple identity properties AND identity renames.
+ * For example, School is a subclass of EducationOrganization which renames educationOrganizationId
+ * to schoolId. An example document identity for a School is { name: schoolId, value: 123 }. The equivalent assignable identity
+ * for this School would be { name: educationOrganizationId, value: 123 }.
  *
- * TODO: Bring this behavior into extractNaturalKey, correctly managing for natural key component ordering.
  */
-export function deriveAssignableFrom(entity: TopLevelEntity, naturalKey: string): AssignableInfo | null {
+export function deriveAssignableFrom(entity: TopLevelEntity, documentIdentity: DocumentIdentity): Assignable | null {
   const { assignableTo }: NullableTopLevelEntity = (entity.data.meadowlark as EntityMeadowlarkData).apiMapping;
   if (assignableTo == null) return null;
   const identityRename: EntityProperty | undefined = entity.identityProperties.find((p) => p.isIdentityRename);
-  if (identityRename == null) return { assignableToName: assignableTo.metaEdName, assignableNaturalKey: naturalKey };
+  if (identityRename == null) return { assignableToName: assignableTo.metaEdName, assignableIdentity: documentIdentity };
 
-  const renamedPropertyName = decapitalize(identityRename.metaEdName);
-  const originalName = decapitalize(identityRename.baseKeyName);
+  const subclassName = decapitalize(identityRename.metaEdName);
+  const superclassName = decapitalize(identityRename.baseKeyName);
+
+  const elementToSubstitute: number = documentIdentity.findIndex((element) => element.name === subclassName);
+  if (elementToSubstitute === -1) return { assignableToName: assignableTo.metaEdName, assignableIdentity: documentIdentity };
+
+  // copy both DocumentIdentity and the individual DocumentElement so original is not mutated
+  const identityCopy: DocumentIdentity = [...documentIdentity];
+  identityCopy[elementToSubstitute] = {
+    ...documentIdentity[elementToSubstitute],
+    name: superclassName,
+  };
 
-  // Substitution regex is lookbehind for "#" and lookahead for first "=" afterward, with renamed property
-  // in between. Using string concatenation to improve readability
-  // eslint-disable-next-line prefer-template
-  const renamedRegex = new RegExp('(?<=#)' + renamedPropertyName + '?(?=\\=)');
   return {
     assignableToName: assignableTo.metaEdName,
-    assignableNaturalKey: naturalKey.replace(renamedRegex, originalName),
+    assignableIdentity: identityCopy,
```

Comment:
Aside from the need to make a copy of the identity before modifying, this is overall simpler after the change as well.